### PR TITLE
New class vpRectOriented and new tools for vpImage added

### PR DESCRIPTION
**Pull request by MarcPouliquenInria**
New class vpRectOriented added: allows to create an oriented recangle (unlike vpRect)
New image tools: mean value, sum of pixels, extraction of oriented rectangular areas, interpolation, correlation
Bug correction:  trying to get min/max values of a vpImage containing no pixels was leading to a segfault

### CODE REVIEW

File: modules/core/src/image/vpImageTools.cpp [hunk 1]
```diff
@@ -339,6 +362,147 @@ void vpImageTools::imageSubtract(const vpImage<unsigned char> &I1, const vpImage
...
+      v1 = (x2 - point.get_i()) * I(x1, y1) + (point.get_i() - x1) * I(x2, y1);
+      v2 = (x2 - point.get_i()) * I(x1, y2) + (point.get_i() - x1) * I(x2, y2);
+    }
+    if (y1 == y2)
+      return v1;
+    return (y2 - point.get_j()) * v1 + (point.get_j() - y1) * v2;
+  }
+  case INTERPOLATION_CUBIC:
+    std::cerr << "Error: bi-cubic interpolation is not implemented." << std::endl;
+    exit(EXIT_FAILURE);
```
Comment (s-trinh) [inline modules/core/src/image/vpImageTools.cpp]: Maybe throwing an exception or just an error message?

File: modules/core/src/image/vpImageTools.cpp [hunk 2]
```diff
@@ -339,6 +362,147 @@ void vpImageTools::imageSubtract(const vpImage<unsigned char> &I1, const vpImage
...
+  default:
+    std::cerr << "Error: invalid interpolation type (" << it << ")" << std::endl;
+    exit(EXIT_FAILURE);
```
Comment (s-trinh) [inline modules/core/src/image/vpImageTools.cpp]: Same comment as above.

File: modules/core/include/visp3/core/vpImage.h [hunk 3]
```diff
@@ -943,6 +949,8 @@ vpImage<Type>::vpImage(vpImage<Type> &&I)
 */
 template <class Type> Type vpImage<Type>::getMaxValue() const
 {
+  if (npixels == 0)
+    throw(vpException(vpException::fatalError, "Cannot compute maximum value of an image containg no pixels"));
```
Comment (fspindle) [inline modules/core/include/visp3/core/vpImage.h]: typo: "Cannot compute maximum value of an empty image"

File: modules/core/include/visp3/core/vpImage.h [hunk 4]
```diff
@@ -951,13 +959,30 @@ template <class Type> Type vpImage<Type>::getMaxValue() const
...
 /*!
   \brief Return the minimum value within the bitmap
 
   \sa getMaxValue()
 */
 template <class Type> Type vpImage<Type>::getMinValue() const
 {
+  if (npixels == 0)
+    throw(vpException(vpException::fatalError, "Cannot compute minimum value of an image containg no pixels"));
   Type m = bitmap[0];
```
Comment (fspindle) [inline modules/core/include/visp3/core/vpImage.h]: typo: "Cannot compute minimum value of an empty image"

File: modules/core/include/visp3/core/vpImage.h [hunk 5]
```diff
@@ -973,6 +998,8 @@ template <class Type> Type vpImage<Type>::getMinValue() const
 */
 template <class Type> void vpImage<Type>::getMinMaxValue(Type &min, Type &max) const
 {
+  if (npixels == 0)
+    throw(vpException(vpException::fatalError, "Cannot compute minimum/maximum values of an image containg no pixels"));
```
Comment (fspindle) [inline modules/core/include/visp3/core/vpImage.h]: typo: "typo: "Cannot compute minimum/maximum values of an empty image""

File: modules/core/include/visp3/core/vpRectOriented.h [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+#include <visp3/core/vpImagePoint.h>
+#include <visp3/core/vpRect.h>
+
+class VISP_EXPORT vpRectOriented
+{
+public:
+  vpRectOriented();
+
+  vpRectOriented(const vpImagePoint center, const double width, const double height, const double theta = 0);
```
Comment (fspindle) [inline modules/core/include/visp3/core/vpRectOriented.h]: Pass by reference : const vpImagePoint &center

File: modules/core/include/visp3/core/vpRectOriented.h [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+
+  vpRectOriented(const vpRect &rect);
+
+  vpRectOriented &operator=(const vpRectOriented &rect);
+
+  vpRectOriented &operator=(const vpRect &rect);
+
+  operator vpRect();
+
+  void setCenter(const vpImagePoint center);
```
Comment (fspindle) [inline modules/core/include/visp3/core/vpRectOriented.h]: Pass arguments by reference

File: modules/core/include/visp3/core/vpRectOriented.h [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+  void setPoints(const vpImagePoint topLeft, const vpImagePoint topRight, const vpImagePoint bottomLeft,
+                 const vpImagePoint bottomRight);
+
```
Comment (fspindle) [inline modules/core/include/visp3/core/vpRectOriented.h]: Pass arguments by reference

File: modules/core/include/visp3/core/vpRectOriented.h [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+  double getWidth() const;
+
+  double getHeight() const;
+
+  void setOrientation(double theta);
+
+  double getOrientation() const;
+
+  bool isInside(vpImagePoint point) const;
```
Comment (fspindle) [inline modules/core/include/visp3/core/vpRectOriented.h]: bool isInside(const vpImagePoint &point) const;

File: modules/core/include/visp3/core/vpRectOriented.h [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+private:
+  vpImagePoint m_center;
+  double m_width;
+  double m_height;
+  double m_theta;
+  vpImagePoint m_topLeft;
+  vpImagePoint m_topRight;
+  vpImagePoint m_bottomLeft;
+  vpImagePoint m_bottomRight;
+  bool isLeft(vpImagePoint pointToTest, vpImagePoint point1, vpImagePoint point2) const;
```
Comment (fspindle) [inline modules/core/include/visp3/core/vpRectOriented.h]: bool isLeft(const vpImagePoint &pointToTest, const vpImagePoint &point1, const vpImagePoint &point2) const;

File: modules/core/include/visp3/core/vpRectOriented.h [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+};
+
+#endif // vpRectOriented_h
```
Comment (fspindle) [inline modules/core/include/visp3/core/vpRectOriented.h]: Add an empty line at the end of the file

File: modules/core/src/image/vpImageTools.cpp [hunk 12]
```diff
@@ -219,6 +218,30 @@ void vpImageTools::imageDifferenceAbsolute(const vpImage<unsigned char> &I1, con
   }
 }
 
+/*!
+  Compute the difference between the two images I1 and I2
+
```
Comment (fspindle) [inline modules/core/src/image/vpImageTools.cpp]: Add a "."

File: modules/core/src/image/vpImageTools.cpp (hunk 1 above)
Comment (fspindle) [reply modules/core/src/image/vpImageTools.cpp]: I would prefer an exception

File: modules/core/src/image/vpImageTools.cpp [hunk 13]
```diff
@@ -339,6 +362,147 @@ void vpImageTools::imageSubtract(const vpImage<unsigned char> &I1, const vpImage
...
+  double x1 = r.getTopLeft().get_i();
+  double y1 = r.getTopLeft().get_j();
+  double t = r.getOrientation();
+  Dst.resize(x_d, y_d);
+  for (unsigned int x = 0; x < x_d; ++x)
+    for (unsigned int y = 0; y < y_d; ++y)
+      Dst(x, y,
+          (unsigned char)interpolate(Src, vpImagePoint(x1 + x * cos(t) + y * sin(t), y1 - x * sin(t) + y * cos(t)),
+                                     vpImageTools::INTERPOLATION_LINEAR));
+}
```
Comment (fspindle) [inline modules/core/src/image/vpImageTools.cpp]: Add here bracket after each for

File: modules/core/src/image/vpImageTools.cpp [hunk 14]
```diff
@@ -339,6 +362,147 @@ void vpImageTools::imageSubtract(const vpImage<unsigned char> &I1, const vpImage
...
+  unsigned int y_d = vpMath::round(r.getWidth());
+  double x1 = r.getTopLeft().get_i();
+  double y1 = r.getTopLeft().get_j();
+  double t = r.getOrientation();
+  Dst.resize(x_d, y_d);
+  for (unsigned int x = 0; x < x_d; ++x)
+    for (unsigned int y = 0; y < y_d; ++y)
+      Dst(x, y, interpolate(Src, vpImagePoint(x1 + x * cos(t) + y * sin(t), y1 - x * sin(t) + y * cos(t)),
+                            vpImageTools::INTERPOLATION_LINEAR));
+}
```
Comment (fspindle) [inline modules/core/src/image/vpImageTools.cpp]: Same add bracket

**Comment by MarcPouliquenInria** [comment]
Corrections done in  [last commit](https://github.com/lagadic/visp/pull/334/commits/a4e03cb3b2524a09701b3c52f97055ecfdfdd31e)

**Comment by fspindle** [comment]
Thanks. I will merge this afternoon

**Pull request merged**